### PR TITLE
We can now store reports done via API

### PR DESCRIPTION
**Pull request by annando**
This is the first step for a better way to report posts. We now store reports that have been done via the API. We don't do anything with them by now.

This PR is done in such an early state since I wanted to layout the general structure of reports so that it can be picked up by someone who is better in UI than me :-)

### CODE REVIEW

File: src/Model/Report.php [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace Friendica\Model;
+
+use Friendica\Core\Logger;
+use Friendica\Database\DBA;
+
+class Report
```
Comment (nupplaphil) [inline src/Model/Report.php]: Can you move this class to `src\Moderation\Repository\Report` and extend it from `BaseRepository`.
So you can make the function like `public function store()` and use `$this->database` instead of all `DBA::` calls.

And then you can pass this `Report` class as an argument at `src\Module\Api\Mastodo\Reports.php`.
There you can copy the behavior of such classes like at the other Mastodon API classes.

If you want, I can make adapt it that way at this PR, but I think this could be a good entrypoint for you at this "new Paradigm" thingy .. but as always just as u like :-)

**Review by nupplaphil — CHANGES_REQUESTED** [review]

File: src/Module/Api/Mastodon/Reports.php [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+use Friendica\Model\Contact;
+use Friendica\Model\Report;
+use Friendica\Module\BaseApi;
+use Friendica\Network\HTTPException;
+
+/**
+ * @see https://docs.joinmastodon.org/methods/accounts/reports/
+ */
+class Reports extends BaseApi
+{
```
Comment (nupplaphil) [inline src/Module/Api/Mastodon/Reports.php:38]: And when you moved it as described above, you can use it like the next two feedbacks I left

```suggestion
{
	/** @var Report */
	protected $report;

	public function __construct(Report $report)
	{
		$this->report = $report;
	}
```

File: src/Module/Api/Mastodon/Reports.php [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+			'comment'    => '',    // Reason for the report (default max 1000 characters)
+			'forward'    => false, // If the account is remote, should the report be forwarded to the remote admin?
+		], $request);
+
+		$contact = Contact::getById($request['account_id'], ['id']);
+		if (empty($contact)) {
+			throw new HTTPException\NotFoundException('Account ' . $request['account_id'] . ' not found');
+		}
+
+		Report::store($contact['id'], $request['status_ids'], $request['comment'], $request['forward'], $uid);
```
Comment (nupplaphil) [inline src/Module/Api/Mastodon/Reports.php]:
```suggestion
		$this->report->store($contact['id'], $request['status_ids'], $request['comment'], $request['forward'], $uid);
```

**Review by nupplaphil — CHANGES_REQUESTED** [review]

**Comment by MrPetovan** [comment]
@nupplaphil I'm planning on doing exactly this next.

**Comment by MrPetovan** [comment]
Here we go.

**Comment by MrPetovan** [comment]
Of course there are 13 seconds between the data method calls and the test method calls. 😖
```
There were 3 failures:

1) Friendica\Test\src\Moderation\Factory\ReportTest::testCreateFromTableRow with data set "default" (array(11, 12, 13, '', false, null), array(), Friendica\Moderation\Entity\Report Object (...))
Failed asserting that 1668301375 matches expected 1668301388.

/woodpecker/src/github.com/friendica/friendica/tests/src/Moderation/Factory/ReportTest.php:88
/woodpecker/src/github.com/friendica/friendica/tests/src/Moderation/Factory/ReportTest.php:99

2) Friendica\Test\src\Moderation\Factory\ReportTest::testCreateFromReportsRequest with data set "default" (12, 13, '', false, array(), Friendica\Moderation\Entity\Report Object (...))
Failed asserting that 1668301375 matches expected 1668301388.

/woodpecker/src/github.com/friendica/friendica/tests/src/Moderation/Factory/ReportTest.php:88
/woodpecker/src/github.com/friendica/friendica/tests/src/Moderation/Factory/ReportTest.php:147

3) Friendica\Test\src\Moderation\Factory\ReportTest::testCreateFromReportsRequest with data set "full" (12, 13, 'Report', true, array(89, 90), Friendica\Moderation\Entity\Report Object (...))
Failed asserting that 1668301375 matches expected 1668301388.

/woodpecker/src/github.com/friendica/friendica/tests/src/Moderation/Factory/ReportTest.php:88
/woodpecker/src/github.com/friendica/friendica/tests/src/Moderation/Factory/ReportTest.php:147
```
Testing `now` is a problem.

File: src/Model/Report.php (hunk 1 above)
Comment (annando) [reply src/Model/Report.php]: Currently I feel like the last tinker on the Titanic. The mail system has to be improved so that moderators can send private messages to non followers, reports need to be transmitted and received from remote systems, a new user class "moderator" needs to be introduced, there is the need for frontend changes to perform reports, the moderator backend has to be done, ... And in addition to all of these, some nasty errors need to be fixed as well.

In short: I have no to little time at the moment to learn something new.

Comment (MrPetovan) [reply src/Model/Report.php]: The moderation overhaul is a large endeavor, please get in touch so that we agree on features before we have any major code down to minimize the potential rewrites.

Comment (annando) [reply src/Model/Report.php]: Definitely. There is a lot of stuff to think about even on the moderation backend, like some protocol, so that every moderator can see what had been done with each report.

Comment (MrPetovan) [reply src/Model/Report.php]: Yes, I'd like use to have a shared design document before we go further with code. This is fine because it is about a Mastodon API endpoint, but anything further requires some sort of plan. What is the most convenient way to do this? Collabora? Google Documents? Also would you like to use instant messages to talk about it?

Comment (annando) [reply src/Model/Report.php]: Surely, we can use Matrix for that as well. My "plan" (I never have a plan, that's why I'm using quotes here) is to start small but then expand without having legacy issues.

Comment (MrPetovan) [reply src/Model/Report.php]: I've registered a Matrix account on an Austrian homeserver and I found two of your accounts and none are the one you have in your profile currently.

**Comment by MrPetovan** [comment]
Now we're good as a base.
